theano/sandbox/mkl/mkl_ndarray.c

```diff
@@ -0,0 +1,1278 @@
+            PyErr_Format(PyExc_RuntimeError,
+                         "MKLNdarray_create_buffer_from_structure: Call dnnAllocateBuffer_F32 failed: %d",
+                         status);
+            return -1;
+        }
+        self->data_size = dnnLayoutGetMemorySize_F32(self->private_layout);
+    }
+
+    return 0;
+}
+
+
+/*
+ * Set user_structure for self.
+ *
+ * nd: number of dimension. nd should <= 16.
+ *
+ * dims: dimension info
+ *
+ */
+int MKLNdarray_set_structure(MKLNdarray *self, int nd, const size_t *dims) {
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_set_structure: input MKLNdarray* self is NULL");
+        return -1;
+    }
+    // assert (self->nd == nd);
+    if (nd > MAX_NDIM) {
+        PyErr_Format(PyExc_ValueError,
+                     "MKLNdarray does not support a %d-dim array. Try array which ndim is <= %d", nd, MAX_NDIM);
+        return -1;
+    }
+
+    self->user_structure[0] = dims[0];
+    self->user_structure[2 * nd - 1] = 1;
+    for (int i = 1; i < nd; i++) {
+        // nchw
+        self->user_structure[i] = dims[i];
+        // chw, hw, w, 1
+        self->user_structure[2 * nd - 1 - i] = self->user_structure[2 * nd - i] * dims[nd - i];
+    }
+
+    return 0;
+}
+
+
+/*
+ * Copy/construct a plain MKLNdarray with dada/structure from a PyArrayObject.
+ *
+ * Need check the dtype and ndim of PyArrayObject.
+ *
+ */
+int MKLNdarray_CopyFromArray(MKLNdarray *self, PyArrayObject *obj) {
+    if (self == NULL || obj == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_CopyFromArray: input self or obj is NULL");
+        return -1;
+    }
+    int ndim = PyArray_NDIM(obj);
+    npy_intp* d = PyArray_DIMS(obj);
+    int typenum = PyArray_TYPE(obj);
+
+    if (typenum != MKL_FLOAT32 && typenum != MKL_FLOAT64) {
+        PyErr_SetString(PyExc_TypeError, "MKLNdarray_CopyFromArray: can only copy from float/double arrays");
+        return -1;
+    }
+
+    if (ndim < 0 || ndim > MAX_NDIM) {
+        PyErr_Format(PyExc_ValueError,
+                     "MKLNdarray does not support a %d-dim array. Try array which ndim is <= %d", ndim, MAX_NDIM);
+        return -1;
+    }
+
+    self->dtype = typenum;
+    self->nd = ndim;
+    size_t dims[MAX_NDIM] = {0};
+    size_t user_size = 1;
+
+    for (int i = 0; i < ndim; i++) {
+        dims[i] = (size_t)d[i];
+        user_size *= dims[i];
+    }
+
+    int err = MKLNdarray_set_structure(self, ndim, dims);
+    if (err < 0) {
+        return err;
+    }
+
+    // prepare user layout and mkl buffer
+    err = MKLNdarray_create_buffer_from_structure(self);
+    if (err < 0) {
+        return err;
+    }
+
+    // copy data to mkl buffer
+    size_t element_size = (size_t)PyArray_ITEMSIZE(obj);
+    // assert (user_size * element_size <= self->data_size);
+    memcpy((void*)self->private_data, (void*)PyArray_DATA(obj), user_size * element_size);
```

Comment:
What if the ndarray is strided? Currently it don't raise an error nor copy correctly in that case.